latch_cli/services/cp/download.py

```diff
@@ -0,0 +1,176 @@
+import time
+from concurrent.futures import ProcessPoolExecutor
+from itertools import repeat
+from pathlib import Path
+from typing import Dict, List, TypedDict
+
+import click
+
+from latch_cli import tinyrequests
+from latch_cli.config.latch import config as latch_config
+from latch_cli.services.cp.config import CPConfig, Progress
+from latch_cli.services.cp.node_utils import LDataNodeType, get_node_data
+from latch_cli.services.cp.path_utils import normalize_path
+from latch_cli.services.cp.progress import ProgressBarManager, ProgressBars
+from latch_cli.services.cp.utils import get_auth_header, human_readable_time, pluralize
+from latch_cli.utils import with_si_suffix
+
+
+class GetSignedUrlData(TypedDict):
+    url: str
+
+
+class GetSignedUrlsRecursiveData(TypedDict):
+    urls: Dict[str, str]
+
+
+def download(
+    src: str,
+    dest: Path,
+    config: CPConfig,
+):
+    normalized = normalize_path(src)
+    node_data = get_node_data(normalized)
+
+    node_has_children = node_data.type in {
+        LDataNodeType.account_root,
+        LDataNodeType.dir,
+        LDataNodeType.mount,
+    }
+
+    if node_has_children:
+        endpoint = latch_config.api.data.get_signed_urls_recursive
+    else:
+        endpoint = latch_config.api.data.get_signed_url
+
+    res = tinyrequests.post(
+        endpoint,
+        headers=get_auth_header(),
+        json={"path": normalized},
+    )
+
+    if res.status_code != 200:
+        raise ValueError(f"Failed to download {src}: {res.json()['error']}")
+
+    json_data = res.json()
+    if node_has_children:
+        dir_data: GetSignedUrlsRecursiveData = json_data["data"]
+
+        if dest.exists() and not dest.is_dir():
+            raise ValueError(f"{str(dest)} is not a directory")
+
+        if dest.exists() and not normalized.endswith("/"):
+            dest = dest / node_data.name
+
+        urls: List[str] = []
+        dests: List[Path] = []
+
+        for rel_path, url in dir_data["urls"].items():
+            urls.append(url)
+            dests.append(dest / rel_path)
```

Comment:
Make a dataclass called download_job or smtg containing the URL and the dest path?